halo2_proofs/tests/frontend_backend_split.rs

```diff
@@ -0,0 +1,612 @@
+                        .expect("todo");
+                    offset += 1;
+                }
+
+                // Enable RLC gate 3 times
+                for abcd in [[3, 5, 3, 5], [8, 9, 8, 9], [111, 222, 111, 222]] {
+                    config.s_rlc.enable(&mut region, offset)?;
+                    let (_, _) = config
+                        .assign_gate(&mut region, &mut offset, None, abcd)
+                        .expect("todo");
+                    let rlc = challenge.map(|ch| {
+                        let [a, b, ..] = abcd;
+                        F::from(a) + ch * F::from(b)
+                    });
+                    region
+                        .assign_advice(|| "", config.e, offset - 1, || rlc)
+                        .expect("todo");
+                    offset += 1;
+                }
+
+                // Enable a dynamic shuffle (sequence from 0 to 15)
+                let table: Vec<_> = (0u64..16).collect();
+                let shuffle = [0u64, 2, 4, 6, 8, 10, 12, 14, 1, 3, 5, 7, 9, 11, 13, 15];
+                assert_eq!(table.len(), shuffle.len());
+
+                for (table_row, shuffle_row) in table.iter().zip(shuffle.iter()) {
+                    region
+                        .assign_fixed(|| "", config.s_shuffle, offset, || Value::known(F::ONE))
+                        .expect("todo");
+                    region
+                        .assign_fixed(|| "", config.s_stable, offset, || Value::known(F::ONE))
+                        .expect("todo");
+                    let shuffle_row0 = Value::known(F::from(*shuffle_row));
+                    region
+                        .assign_advice(|| "", config.a, offset, || shuffle_row0)
+                        .expect("todo");
+                    let table_row0 = Value::known(F::from(*table_row));
+                    region
+                        .assign_advice(|| "", config.b, offset, || table_row0)
+                        .expect("todo");
+                    offset += 1;
+                }
+
+                Ok((offset, instance_copy))
+            },
+        )?;
+
+        Ok((rows, instance_copy))
+    }
+}
+
+impl<F: Field + From<u64>, const WIDTH_FACTOR: usize> Circuit<F> for MyCircuit<F, WIDTH_FACTOR> {
+    type Config = Vec<MyCircuitConfig>;
+    type FloorPlanner = SimpleFloorPlanner;
+    #[cfg(feature = "circuit-params")]
+    type Params = ();
+
+    fn without_witnesses(&self) -> Self {
+        self.clone()
+    }
+
+    fn configure(meta: &mut ConstraintSystem<F>) -> Vec<MyCircuitConfig> {
+        assert!(WIDTH_FACTOR > 0);
+        (0..WIDTH_FACTOR)
+            .map(|_| Self::configure_single(meta))
+            .collect()
+    }
+
+    fn synthesize(
+        &self,
+        config: Vec<MyCircuitConfig>,
+        mut layouter: impl Layouter<F>,
+    ) -> Result<(), Error> {
+        // 2 queries from first gate, 3 for permutation argument, 1 for multipoen, 1 for off-by-one
+        //   errors, 1 for off-by-two errors?
```

Comment:
> 1 for off-by-two errors

It should be 1 extra row for the last row of grand product poly to check if it products to `1`.

Reply:
Thanks for clarifying this point!

Reply:
Resolved in https://github.com/privacy-scaling-explorations/halo2/pull/254/commits/134143c618f41aea4805fd0cff14962c819397f1